### ISSUE TITLE
"Pick"/"keyof" projection giving spurious errors when "--strictNullChecks" is not used.

### ISSUE THREAD

**Issue by cakoose** [Microsoft/TypeScript#18833]
**TypeScript Version:**  2.5.3

**Code**

```ts
type User = {
    id: number,
    name: string,
    age: number,
}

type UserMinimumFields = 'id' | 'name';

// Returns a User object.  The 'id' and 'name' fields are always returned.
// You may also specify extra fields you want returned.
declare function getUser<F extends keyof User>(extraFields: F[]): Pick<User,F | UserMinimumFields>;

function f() {
    const u1 = getUser([]);
    u1.id;   // spurious error
    u1.name; // spurious error
    u1.age;  // legit error
    u1.blah; // legit error

    const u2 = getUser(['age']);
    u2.id;
    u2.name;
    u2.age;
    u2.blah; // legit error

    const u3 = getUser(['age', 'name']);
    u2.id;
    u2.name;
    u2.age;
    u2.blah; // legit error

    getUser(['name', 'blah']); // error
}
```

**Expected behavior:**

With "--structNullChecks", "tsc" only reports errors on the lines marked "legit error".  This is what I expect.

```
$ ./node_modules/.bin/tsc --strictNullChecks typescript.ts 
typescript.ts(17,8): error TS2339: Property 'age' does not exist on type 'Pick<User, UserMinimumFields>'.
typescript.ts(18,8): error TS2339: Property 'blah' does not exist on type 'Pick<User, UserMinimumFields>'.
typescript.ts(24,8): error TS2339: Property 'blah' does not exist on type 'Pick<User, "id" | "name" | "age">'.
typescript.ts(30,8): error TS2339: Property 'blah' does not exist on type 'Pick<User, "id" | "name" | "age">'.
typescript.ts(32,13): error TS2345: Argument of type '("name" | "blah")[]' is not assignable to parameter of type '("id" | "name" | "age")[]'.
  Type '"name" | "blah"' is not assignable to type '"id" | "name" | "age"'.
    Type '"blah"' is not assignable to type '"id" | "name" | "age"'.
```

**Actual behavior:**

Without "--strictNullChecks", "tsc" reports all the same errors as above, but also reports the lines marked "spurious error".

```
$ ./node_modules/.bin/tsc typescript.ts 
typescript.ts(15,8): error TS2339: Property 'id' does not exist on type 'Pick<User, any>'.
typescript.ts(16,8): error TS2339: Property 'name' does not exist on type 'Pick<User, any>'.
typescript.ts(17,8): error TS2339: Property 'age' does not exist on type 'Pick<User, any>'.
typescript.ts(18,8): error TS2339: Property 'blah' does not exist on type 'Pick<User, any>'.
typescript.ts(24,8): error TS2339: Property 'blah' does not exist on type 'Pick<User, "id" | "name" | "age">'.
typescript.ts(30,8): error TS2339: Property 'blah' does not exist on type 'Pick<User, "id" | "name" | "age">'.
typescript.ts(32,13): error TS2345: Argument of type '("name" | "blah")[]' is not assignable to parameter of type '("id" | "name" | "age")[]'.
  Type '"name" | "blah"' is not assignable to type '"id" | "name" | "age"'.
    Type '"blah"' is not assignable to type '"id" | "name" | "age"'.
```

I thought maybe it might be a bug because I couldn't see why "--strictNullChecks" should make a difference here.  (Apologies if this is actually expected behavior.)

**Comment by DanielRosenwasser** [comment]
Seems like this is pretty undesirable. Maybe we shouldn't make inferences from the `undefinedWideningType`

**Comment by ahejlsberg** [comment]
The core issue here is that we use different types for the expression `[]` in default type checking mode (where it is `any[]`) and strict null checking mode (where it is `never[]`), and for reasons of backward compatibility we can't change that. I will think about other possible ways we could make your example behave the same in both modes, but meanwhile a simple workaround is to change `getUser` to use an intersection in the return type:

```ts
declare function getUser<F extends keyof User>(extraFields: F[]): Pick<User, F> & Pick<User, UserMinimumFields>;
```

**Comment by ahejlsberg** [comment]
This issue was fixed by #19185. Without `--strictNullChecks` all of the property accesses involving `u1` now succeed because the inferred type of `u1` is `{ [x: string]: any }`. This is less restrictive than before, but I think it is the right behavior given that we consider `[]` to be of type `any[]`.
